Sources/ClientRuntime/Networking/Http/HTTPResponseClosure.swift

```diff
@@ -7,10 +7,32 @@
 
 public typealias HTTPResponseClosure<OperationStackOutput> = (HttpResponse) async throws -> OperationStackOutput
 
+public typealias HTTPResponseOutputBinding<OperationStackOutput, Reader> =
+    (HttpResponse, Reader) async throws -> OperationStackOutput
+
+/// Provides a response closure for a type that conforms to `HttpResponseBinding`.
+///
+/// This allows for use of JSON and FormURL serialized types with closure-based deserialization.
```

Comment:
SmithyReadWrite is intended to be a container for things that are shared across serialization formats (i.e. XML, FormURL, JSON).  That is why the `Reader` type in `DocumentReadingClosure` is generic above: in the future, it might not be an XML reader.

To make the other serialization formats "closure-based" in the sense that XML is now, we will need to:
- Provide a `Writer` / `Reader` for that format like the ones I provide for XML here.
- Switch the serialization codegen over to the same serialization as XML.  It would use the same codegen but not use the XML-specific features like namespace, attributes, etc.
- Delete a LOT of codegen code, since all serialization formats will use roughly the same codegen.